src/ibus.h

```diff
@@ -46,7 +46,6 @@
 #include <ibuskeysyms.h>
 #include <ibuskeymap.h>
 #include <ibuskeys.h>
-#include <ibusenumtypes.h>
```

Comment:
I think deleting ibusenumtypes.h in ibus.h would break the compatibility. Users include ibus.h only at present.
I guess we need another solution.

Reply:
I'm open to ideas.

I tried many things to break the dependency cycle before I arrived at this solution.

Reply:
How about moving ibus.h from $(ibus_headers)? likes this?:
```
diff --git a/src/Makefile.am b/src/Makefile.am
index adc7f04a..dd5aacc1 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -127,7 +127,6 @@ ibus_resources_sources =	\
     ibusresources.h	     \
     $(NULL)
 ibus_headers =		     \
-    ibus.h		     \
     ibusaccelgroup.h	     \
     ibusattribute.h	     \
     ibusattrlist.h	     \
@@ -169,6 +168,7 @@ ibus_headers =		\
 ibusincludedir = $(includedir)/ibus-@IBUS_API_VERSION@
 ibus_public_headers =	     \
     $(ibus_headers)	     \
+    ibus.h		     \
     ibusenumtypes.h	     \
     ibusversion.h	     \
     $(NULL)
@@ -213,6 +213,7 @@ if HAVE_INTROSPECTION
 introspection_sources =	 \
     $(ibus_headers)		 \
     $(ibus_sources)		 \
+    ibus.h			 \
     $(NULL)
 built_introspection_sources =	 \
     $(ibus_enumtypes_sources)	 \
```

Reply:
No, unfortunately this doesn't work:

````
In file included from gencomposetable.c:23:
./ibus.h:49:10: fatal error: ibusenumtypes.h: No such file or directory
   49 | #include <ibusenumtypes.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
````

Can you not reproduce the failure locally? For me it is as simple as

````
VALAC=/usr/bin/valac-0.56 VAPIGEN=/usr/bin/vapigen-0.56 ./autogen.sh --with-ucd-dir=/usr/share/unicode-data
make -C src ibusresources.c -j8
````

`ibusresources.c` and `ibusenumtypes.h` are both in `BUILT_SOURCES`, and while there's a "barrier" between `BUILT_SOURCES` and the rest of the build (that is, all files in `BUILT_SOURCES` are built before `make` begins building anything else), I don't know of any way to specify a dependency *between* two targets in `BUILT_SOURCES`.

Reply:
Oh! Maybe I can use automake's `EXTRA_target_DEPENDENCIES` (see https://www.gnu.org/software/automake/manual/html_node/Program-and-Library-Variables.html).

Since it's a source file in `gen-internal-compose-table` that fails to compile, I can add

````
EXTRA_gen_internal_compose_table_DEPENDENCIES = ibusenumtypes.h
````

Let me try this...

Reply:
Damn, that doesn't work.

I believe it does not work because it adds a dependency on the final `gen-internal-compose-table` binary, and not on the `.o` file that is being compiled when the error occurs (`gen_internal_compose_table-gencomposetable.o`).

I can see that `ibusenumtypes.h` is being generated now, but not as a dependency of `gen_internal_compose_table-gencomposetable.o`.


````
mattst88@framework ibus % make -C src ibusresources.c -j8
make: Entering directory '/home/mattst88/projects/ibus/src'
  GEN      ibus.gresources.xml
  CC       gen_internal_compose_table-gencomposetable.o
  CC       gen_internal_compose_table-ibuscomposetable.o
  CC       gen_internal_compose_table-ibuserror.o
  CC       gen_internal_compose_table-ibuskeynames.o
  CC       gen_internal_compose_table-ibuskeyuni.o
  GEN      ibusenumtypes.h
In file included from gencomposetable.c:23:
./ibus.h:49:10: fatal error: ibusenumtypes.h: No such file or directory
   49 | #include <ibusenumtypes.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
````

If there is a way to make an `EXTRA_*_DEPENDENCIES` variable for a `.o` file, I do not know how to do it.